pkg/util/bytesbufpool/bytesbufpool.go

```diff
@@ -0,0 +1,58 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bytesbufpool
+
+import (
+	"bytes"
+	"sync"
+)
+
+var pool = sync.Pool{}
+
+const defaultCap = 100
+
+// Get a buffer with default capacity.
+func Get() *bytes.Buffer {
+	return GetWithCap(defaultCap)
+}
+
+// GetWithCap get a buffer with at least n bytes can be written to it without another allocation.
+// If n is negative, GetWithCap will panic.
+func GetWithCap(n int) *bytes.Buffer {
+	val := pool.Get()
+	if val != nil {
+		buf := val.(*bytes.Buffer)
+		if buf.Cap() < n {
+			buf.Grow(n)
+		}
+		return buf
+	}
+	return bytes.NewBuffer(make([]byte, 0, n))
+}
+
+func Put(buf *bytes.Buffer) {
```

Comment:
This effectively creates a global/central lock around (potentially) all buffer allocations, right? Do we understand the implications of that?

Reply:
Thanks for the review.

Yes, but I haven't done it this way before. Could you explain some of the potential impacts?

Reply:
Oh, nevermind, seems like it's more of a ring-buffer, so this objection is invalid.